Random Groups:
--

1. "James H", "Caitlin", "Max"
2. "Amy", "Lauren", "Olsi"
3. "Jacob", "Flavio", "Alexis"
4. "Cameron", "Sean", "Michael" 
5. "Chase", "Samuel", "James S"
6. "Jennifer", "Luis", "Killian"

https://student.desmos.com/?prepopulateCode=gsbefn

---
# Section 1.3: Triangular Systems
---

Let $L \in \mathbb{R}^{n \times n}$ be a **lower-triangular matrix** and $U \in \mathbb{R}^{n \times n}$ be an **upper-triangular matrix**. Then,

$$
L = 
\begin{bmatrix}
l_{11} \\
\vdots & \ddots \\
l_{n1} & \cdots & l_{nn}
\end{bmatrix}
\qquad \text{and} \qquad
U = 
\begin{bmatrix}
u_{11} & \cdots & u_{1n} \\
& \ddots & \vdots \\
& & u_{nn}
\end{bmatrix}.
$$

> ### Theorem:
> Let $G$ be an $n \times n$ triangular matrix. Then
> 
> $$ 
\text{$G$ is nonsingular} \quad \Longleftrightarrow \quad g_{ii} \neq 0, \quad i = 1,\ldots,n.
$$

---
### Exercise: 

Prove the above theorem.

### Proof:

Since $G$ is triangular, the determinant of $G$ is the product of its diagonal entries; that is, $\det(G) = g_{11} g_{22} \cdots g_{nn}$. Thus, $G$ is nonsingular if and only if $g_{11} g_{22} \cdots g_{nn} \ne 0$, which holds if and only if $g_{ii} \ne 0$ for $i = 1,\ldots,n$. $\blacksquare$

### Proof:

We will prove the theorem by mathematical induction on $n$. 

First let $n = 1$ and let $G$ be an $n \times n$ triangular matrix. Then $G = \begin{bmatrix} g_{11} \end{bmatrix}$. 

Suppose $G$ is nonsingular; that is, the only solution of $Gx = 0$ is $x = 0$. Suppose, for the sake of contradiction, that $g_{11} = 0$. Then, letting $x = \begin{bmatrix} 1 \end{bmatrix}$, we have that $Gx = \begin{bmatrix} g_{11} x_1 \end{bmatrix} = \begin{bmatrix} 0 \cdot 1 \end{bmatrix} = \begin{bmatrix} 0 \end{bmatrix} = 0$; but $x \ne 0$, contradicting the assumption that $G$ is nonsingular. Therefore, we must have $g_{11} \ne 0$.

Now, suppose that $g_{11} \ne 0$. Suppose that $Gx = 0$. Then, $g_{11} x_1 = 0$; dividing both sides by $g_{11}$ (which we can do since $g_{11} \ne 0$), we conclude that $x_1 = 0$. Thus, $x = \begin{bmatrix} x_1 \end{bmatrix} = 0$. 

Next, we assume that the theorem is true for $n = k$. L

---

### Exercise:
Solve the following lower-triangular system $Lx = b$ and use `Julia` to check your answer.

$$
\begin{bmatrix}
2 & 0 & 0 & 0 \\
-1 & 2 & 0 & 0 \\
3 & 1 & -1 & 0 \\
4 & 1 & -3 & 3
\end{bmatrix}
\begin{bmatrix}
x_1 \\ x_2 \\ x_3 \\ x_4
\end{bmatrix} =
\begin{bmatrix}
2 \\ 3 \\ 2 \\ 9
\end{bmatrix}.
$$

In [1]:
L = [2 0 0 0; -1 2 0 0; 3 1 -1 0; 4 1 -3 3.0]
b = [2, 3, 2, 9.0]

x = L\b

4-element Array{Float64,1}:
 1.0
 2.0
 3.0
 4.0

---

In general, the solution of a lower-triangular system $Lx = b$ is given by

$$
x_i = \left( b_i - \sum_{j=1}^{i-1} l_{ij}x_j \right) \bigg/ l_{ii}, \quad i = 1, \ldots, n.
$$

---

### Exercise:

Write a `Julia` function, `x = row_fs(L, b)`, that returns the solution $x$ of a lower-triangular system $Lx = b$ using _forward substitution_. Make sure to include a check that the matrix $L$ is nonsingular. Test your code for accuracy and efficiency.

In [2]:
using LinearAlgebra

In [3]:
function row_fs(L, b)
    
    if !istril(L)
        error("L is not lower triangular")
    end
    
    n = size(L,1)
    
    x = copy(b) # OR: x = zeros(n)
    
    for i = 1:n
        # x[i] = b[i] # if x initialized as zeros(n)
        for j = 1:i-1
            x[i] -= L[i,j]*x[j]
        end
        if L[i,i] == 0.0
            error("L is singular")
        end
        x[i] /= L[i,i]
    end
    
    return x
end

row_fs (generic function with 1 method)

In [4]:
L = [2 0 0 0; -1 2 0 0; 3 1 -1 0; 4 1 -3 3.0]
b = [2, 3, 2, 9.0]

x = row_fs(L, b)

4-element Array{Float64,1}:
 1.0
 2.0
 3.0
 4.0

In [5]:
typeof(L)

Array{Float64,2}

In [6]:
istril(L)

true

In [7]:
methods(istril)

# 14 methods for generic function "istril":
[1] istril(A::BitArray{2}) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/bitarray.jl:165
[2] istril(M::SymTridiagonal) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/tridiag.jl:274
[3] istril(M::Tridiagonal) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/tridiag.jl:575
[4] istril(A::LowerTriangular) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl:281
[5] istril(A::UnitLowerTriangular) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl:282
[6] istril(A::Adjoint) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl:285
[7] istril(A::Transpose) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl:286
[8] istril(D::Diagonal) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/diagonal.jl:131
[9] istril(M::Bidiagonal) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/bidiag.jl:250
[10] istril(A::SparseArrays.SparseMatrixCSC) in SparseArrays at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/SparseArrays/src/sparsematrix.jl:3234
[11] istril(A::AbstractArray{T,2} where T) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/generic.jl:1162
[12] istril(A::AbstractArray{T,2} where T, k::Integer) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/generic.jl:1162
[13] istril(x::Number) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/generic.jl:1171
[14] istril(::UniformScaling) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/uniformscaling.jl:104

In [8]:
@which istril(L)

istril(A::AbstractArray{T,2} where T) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/generic.jl:1162

---
If you open the file `julia/stdlib/v1.2/LinearAlgebra/src/generic.jl` and go to line 1162, you will see

```julia
function istril(A::AbstractMatrix, k::Integer = 0)
    require_one_based_indexing(A)
    m, n = size(A)
    for j in max(1, k + 2):n
        for i in 1:min(j - k - 1, m)
            iszero(A[i, j]) || return false
        end
    end
    return true
end
istril(x::Number) = true
```
---

In [9]:
@edit istril(L)

In [17]:
n = 5
L = LowerTriangular(rand(n,n))

5×5 LowerTriangular{Float64,Array{Float64,2}}:
 0.251022   ⋅         ⋅         ⋅         ⋅      
 0.216567  0.190719   ⋅         ⋅         ⋅      
 0.105814  0.5727    0.972102   ⋅         ⋅      
 0.998731  0.30645   0.185685  0.108457   ⋅      
 0.498394  0.888728  0.217014  0.721708  0.092518

In [18]:
L.data

5×5 Array{Float64,2}:
 0.251022  0.295733  0.479975  0.610201  0.710456
 0.216567  0.190719  0.754775  0.932257  0.936163
 0.105814  0.5727    0.972102  0.730147  0.161495
 0.998731  0.30645   0.185685  0.108457  0.530526
 0.498394  0.888728  0.217014  0.721708  0.092518

In [19]:
L[1,2]

0.0

In [20]:
typeof(L)

LowerTriangular{Float64,Array{Float64,2}}

In [21]:
@which istril(L)

istril(A::LowerTriangular) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl:281

---
If you open the file `julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl` and go to line 281, you will see

```julia
istril(A::LowerTriangular) = true
istril(A::UnitLowerTriangular) = true
istriu(A::UpperTriangular) = true
istriu(A::UnitUpperTriangular) = true
istril(A::Adjoint) = istriu(A.parent)
istril(A::Transpose) = istriu(A.parent)
istriu(A::Adjoint) = istril(A.parent)
istriu(A::Transpose) = istril(A.parent)
```
---

In [15]:
L'

5×5 Adjoint{Float64,LowerTriangular{Float64,Array{Float64,2}}}:
 0.642885  0.741454  0.70927   0.0147822  0.269491 
 0.0       0.575002  0.714529  0.260178   0.0455492
 0.0       0.0       0.623339  0.228832   0.69635  
 0.0       0.0       0.0       0.368133   0.809246 
 0.0       0.0       0.0       0.0        0.446726 

In [16]:
L = tril(rand(n,n))

5×5 Array{Float64,2}:
 0.138224  0.0       0.0       0.0       0.0      
 0.871117  0.467573  0.0       0.0       0.0      
 0.734591  0.460635  0.46425   0.0       0.0      
 0.964278  0.854809  0.278847  0.876663  0.0      
 0.363064  0.403454  0.595334  0.876801  0.0270823

---

## Column-oriented forward substitution

Notice that the algorithm above is **row-oriented**, which is good for computer languages that store matrices in row-major order (like C/C++ or Python), but is bad for column-major order languages (like Fortran, MATLAB, or Julia).

We can produce a forward-substitution algorithm that is **column-oriented** by using the concept of **block-matrix multiplication**.

---

### Exercise:

Let the lower-triangular linear system $Lx = b$ be partitioned as

$$
\begin{bmatrix}
l_{11} & 0 \\
\hat{l} & \hat{L} \\
\end{bmatrix}
\begin{bmatrix}
x_1 \\ \hat{x}
\end{bmatrix} = 
\begin{bmatrix}
b_1 \\ \hat{b}
\end{bmatrix}
$$

where $\hat{L}$ is an $(n-1) \times (n-1)$ *lower-triangular* matrix and $\hat{x}$ and $\hat{b}$ are vectors of length $n - 1$.

1. Perform block-matrix multiplication.

2. Determine a _recursive algorithm_ for solving the linear system $Lx = b$.

3. Implement your recursive algorithm in `Julia` (name your function `col_rfs`).

4. Test your code for accuracy and efficiency.

---

### Exercise:

Use column-oriented forward substitution to solve the lower-triangular system.

$$
\begin{bmatrix}
2 & 0 & 0 & 0 \\
-1 & 2 & 0 & 0 \\
3 & 1 & -1 & 0 \\
4 & 1 & -3 & 3
\end{bmatrix}
\begin{bmatrix}
x_1 \\ x_2 \\ x_3 \\ x_4
\end{bmatrix} =
\begin{bmatrix}
2 \\ 3 \\ 2 \\ 9
\end{bmatrix}
$$


---

### Exercise:

Inspired by your calculations in the previous exercise, write a non-recursive column-oriented forward-substitution code in `Julia` (name your function `col_fs`). Test your function for accuracy and efficiency.

---

### Exercise:

How do your `row_fs` and `col_fs` functions compare to `Julia`'s built-in `x = L\b`?

---

### Exercise:

Determine the total number of flops to solve an $n \times n$ triangular linear system using two different methods:

1. Directly count the flops in your nonrecursive code.
2. Count the flops using the recursive version of your code.

---

### Exercise:

Write a `Julia` function `col_fs!(b, L)` that **overwrites** the array `b` with the solution of the linear sytem $Lx = b$. Test the efficiency of your function.

---